src/librustc_data_structures/macros.rs

```diff
@@ -16,6 +16,6 @@ macro_rules! static_assert {
         // Use the bool to access an array such that if the bool is false, the access
         // is out-of-bounds.
         #[allow(dead_code)]
-        static $name: () = [()][!$test as usize];
+        static $name: () = [()][!($test: bool) as usize];
```

Comment:
You could use `#[allow_internal_unstable]` instead, maybe?
Also, has the `const _ = ...;` proposal been implemented?

Reply:
> Also, has the const _ = ...; proposal been implemented?

It has been implemented, but it does not cope well with `allow_internal_unstable`, because the crate using `static_assert` won't have the feature for `_` items enabled, thus causing an error about item with name `_` being declared multiple times

Reply:
I'm very confused, how could that even happen?!

Reply:
I don't know, I presume it's because we now do have multiple items with the name `_`, but the multiple items with same name check doesn't know about `allow_internal_unstable` and thus assumes the relevant feature gate (for `_` names) isn't active.

Reply:
But the implementation should be using `gensym` so that should be impossible?

Reply:
It does. https://github.com/rust-lang/rust/blob/406cbf1a39c99ef7b7f329a44f1d1c7af7a8ecda/src/libsyntax/parse/parser.rs#L6350-L6353

That PR doesn't quite do the right thing though. It accidentally stabilized `static _` while `const _` is feature gated (https://play.rust-lang.org/?version=beta&mode=debug&edition=2015&gist=192ade120e9d1b4c0ff9256c50cdd649).

I presume something is going wrong with hygiene. `macro_rules` macros should not be affected, but maybe the hash or eq impls of `Ident` are wrong? The code checking for naming collisions uses a `FxHashMap`.

Reply:
> maybe the hash or eq impls of Ident are wrong?

nope, that goes through `Symbol` which is just a `u32` id.

Reply:
This only happens cross-crate. Within the `rustc_data_structures` there's no problems (which also explains why my playground tests failed to repro)

Reply:
Gensyms (and other hygiene data) are not kept in metadata and are therefore lost in cross-crate scenarios.
That's an old and fundamental issue that's larger than both `use X as _` and `const _` and e.g. prevents stabilization of macros 2.0 as well.

Reply:
Oh, is `pub use foo::{Bar as _, Baz as _};` broken as well?
I thought we excluded those from reexports, or maybe I just hit that bug with uniform paths and special-cased the canaries instead of `_`?

Reply:
>I thought we excluded those from reexports

That may be the case (or maybe that's the case only for canaries, I don't remember), but strictly speaking this is not a correct thing to do for non-canaries.
`pub use foo::{MyTrait1 as _, MyTrait2 as _}` should indeed bring both traits' methods in scope even cross-crate

Reply:
But you would only be able to get those in scope with a glob import? That makes sense, I guess. We could check for `_` and have them in a separate list that gets re-gensym'd in the downstream crate, but that seems like a hack.

Reply:
> Gensyms (and other hygiene data) are not kept in metadata and are therefore lost in cross-crate scenarios.

couldn't we encode symbols as `enum MabyeGenSym { Gensym(Symbol), Normal(Symbol) }` to "preserve" this by regensymming during decoding in the `Gensym` variant?

Reply:
@oli-obk No, because that won't preserve the association. Anyway, gensyms are just an implementation hack, you don't actually need a name for these items.

Reply:
I can refactor AST and HIR to have `Option<Symbol>` for their names

Reply:
That seems overkill for now, and also, we can move this into an issue and land a simpler version of this patch.

Reply:
by simpler you mean keeping the actual names for the assert items?

Reply:
yeah